### PR TITLE
Consolidate token list controller data

### PR DESCRIPTION
**Pull request by Cal-L**
### **Issue**
- `TokenListController` should consolidate all tokens regardless of static or dynamic into one data type: `TokenListToken`
- For images, `TokenListController` returns a file name from static tokens and URLs from dynamic tokens. 
- FUTURE PROPOSAL: `TokenListController` should split the images field into two separate properties `iconPath` for static tokens and `iconUrl` for dynamic tokens.

### **Resolves/Changes**
- Map static tokens to return `TokenListToken` type. The idea was already there but I mapped it more, added `address`, `occurrences`, and `aggregators`. 
- Export `TokenListToken` type from `TokenListController` for use in the apps
- Update `TokenListController` tests

### CODE REVIEW

File: src/assets/TokenListController.test.ts [hunk 1]
```diff
@@ -11,17 +11,29 @@ import {
...
 const TOKEN_END_POINT_API = 'https://token-api.airswap-prod.codefi.network';
 const timestamp = Date.now();
 
-const staticTokenList: any = {};
-for (const tokenAddress in contractmap) {
-  const { erc20, logo, ...token } = contractmap[tokenAddress];
+const staticTokenList: TokenListMap = {};
+for (const tokenAddress in contractMap) {
+  const { erc20, logo: filePath, ...token } = contractMap[tokenAddress];
+  const extType = filePath.split('.')[1].toUpperCase() as MediaExtType;
```
Comment (rickycodes) [inline src/assets/TokenListController.test.ts:26]: this isn't going to work with filenames with more than just the extension (i know we don't have that here) but `my.cool.token.jpg` would break here:

image

additionally it's just slow to use `.split` when all you really want is a string anyway. I think using substring methods is about 80% faster. you probably want:
```
filePath.substr(filePath.lastIndexOf('.') + 1)
```

File: src/assets/TokenListController.ts [hunk 2]
```diff
@@ -195,11 +225,21 @@ export class TokenListController extends BaseController<
    * Fetching token list from the contract-metadata as a fallback
    */
   async fetchFromStaticTokenList(): Promise<void> {
-    const tokenList: TokenMap = {};
-    for (const tokenAddress in contractmap) {
-      const { erc20, logo, ...token } = contractmap[tokenAddress];
+    const tokenList: TokenListMap = {};
+    for (const tokenAddress in contractMap) {
+      const { erc20, logo: filePath, ...token } = contractMap[tokenAddress];
+      const extType = filePath.split('.')[1].toUpperCase() as MediaExtType;
```
Comment (rickycodes) [inline src/assets/TokenListController.ts:231]: same as above

**Review by rickycodes — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -37,7 +37,7 @@
   "dependencies": {
     "@ethereumjs/common": "^2.3.1",
     "@ethereumjs/tx": "^3.2.1",
-    "@metamask/contract-metadata": "^1.28.0",
+    "@metamask/contract-metadata": "1.28.0",
```
Comment (rickycodes) [inline package.json:40]: why the change?

**Review by rickycodes — APPROVED** [review]
on small nit re: pinning dep, but otherwise LGTM

File: src/apis/token-service.ts [hunk 4]
```diff
@@ -12,6 +12,10 @@ function getTokenMetadataURL(chainId: string, tokenAddress: string) {
   return `${END_POINT}/tokens/${chainId}?address=${tokenAddress}`;
 }
 
+// Token list averages 1.6 MB in size
```
[the remark this answers is not in the document]
Comment (Cal-L) [reply src/apis/token-service.ts:15]: Yep 😬. By default, once every hour.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/apis/token-service.ts:15]: Welp! Not something we should address in this PR, but we'll have to revisit this. That is too much network traffic.

File: src/assets/TokenListController.ts [hunk 5]
```diff
@@ -17,30 +17,53 @@ const DEFAULT_THRESHOLD = 60 * 30 * 1000;
...
+  filePath: string;
+  type: MediaExtType;
+};
+
+export type TokenListToken = {
   address: string;
-  decimals: number;
-  symbol: string;
+  iconUrl: string | null;
+  iconPath: IconPath | null;
```
Comment (Gudahtt) [inline src/assets/TokenListController.ts:56]: What is the purpose of this property? I see that it has two properties ("filePath" which is the _filename_, not really a path, and "type" which is the file extension), but I don't know what it's for. Especially if `iconUrl` will remain.

I had thought your intention was to normalize `iconUrl` to be a real URL for both static and dynamic tokens, but that no longer appears to be your goal.

File: src/assets/TokenListController.ts [hunk 6]
```diff
@@ -17,30 +17,53 @@ const DEFAULT_THRESHOLD = 60 * 30 * 1000;
...
+export type IconPath = {
+  filePath: string;
+  type: MediaExtType;
+};
+
+export type TokenListToken = {
   address: string;
-  decimals: number;
-  symbol: string;
+  iconUrl: string | null;
```
Comment (Gudahtt) [inline src/assets/TokenListController.ts:55]: Is it possible for this to be `null`? It seems like it is always set.

File: src/assets/TokenListController.ts (hunk 5 above)
Comment (Cal-L) [reply src/assets/TokenListController.ts]: Removed iconPath

File: src/assets/TokenListController.ts [hunk 7]
```diff
@@ -257,8 +318,8 @@ export class TokenListController extends BaseController<
    *  otherwise a call to the API service will be made.
    * @returns Promise that resolves into a TokenList
    */
-  async fetchFromCache(): Promise<Token[]> {
-    const { tokensChainsCache, ...tokensData }: TokenListState = this.state;
+  async fetchFromCache(): Promise<TokenListToken[] | null> {
```
[the remark this answers is not in the document]
Comment (Cal-L) [reply src/assets/TokenListController.ts:310]: Updated description

File: src/assets/TokenListController.ts [hunk 8]
```diff
@@ -195,11 +219,20 @@ export class TokenListController extends BaseController<
...
-    for (const tokenAddress in contractmap) {
-      const { erc20, logo, ...token } = contractmap[tokenAddress];
+    const tokenList: TokenListMap = {};
+    for (const tokenAddress in contractMap) {
+      const { erc20, logo: filePath, ...token } = (contractMap as ContractMap)[
+        tokenAddress
+      ];
       if (erc20) {
-        tokenList[tokenAddress] = { ...token, iconUrl: logo };
+        // Specify iconUrl here as filePath for backwards compatibility for extension
```
Comment (Gudahtt) [inline src/assets/TokenListController.ts:228]: Nit: this comment was meant to explain why both `iconUrl` and `iconPath` were set, but now that `iconPath` is gone it is no longer necessary.

File: src/assets/TokenListController.ts [hunk 9]
```diff
@@ -17,30 +17,47 @@ const DEFAULT_THRESHOLD = 60 * 30 * 1000;
...
+};
+
+export type DynamicToken = {
+  address: string;
+  occurrences: number;
+  aggregators: string[];
+  iconUrl: string;
+} & BaseToken;
+
+export enum MediaExtType {
```
Comment (Gudahtt) [inline src/assets/TokenListController.ts:42]: This type appears to be unused

**Review by Gudahtt — COMMENTED** [review]
I noticed two minor issues, but this looks good once the extraneous stuff is removed!

**Review by Gudahtt — APPROVED** [review]
LGTM!
